TodoApp.ConsoleApp/UI/UserInput.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            Console.WriteLine(question);
+            string text = "";
+            if (required)
+            {
+                text = ReadInput(required);
+
+                if (string.IsNullOrEmpty(text))
+                {
+                    text = defaultValue;
+                }
+            }
+
+            else
+            {
+                text = defaultValue;
+            }
+
+            return text;
+        }
+
+        public static int ReadOption(string question, string[] availableOptions, bool required = false, int defaultValue = 0)
+        {
+            Console.WriteLine(question);
+            int index = 1;
+            foreach (var option in availableOptions)
+            {
+                Console.WriteLine($"\t{index}: {option}");
+                index++;
+            }
+
+            index = 1;
+            bool isValid = false;
+            string input = "";
+
+            if (required)
+            {
+                do
+                {
+                    input = ReadInput(required);
+                    isValid = int.TryParse(input, out index) && (index > 0 && index <= availableOptions.Length);
+
+                    if (!isValid)
+                    {
+                        Console.WriteLine($"Please, enter a valid number between 1 and {availableOptions.Length}!");
+                    }
+                }
+                while (!isValid);
+            }
+
+            else
+            {
+                input = ReadInput(required);
+                isValid = int.TryParse(input, out index) && (index > 0 && index < availableOptions.Length);
+                if (!isValid)
+                {
+                    index = defaultValue;
+                }
+            }
+
+            return index;
+        }
+
+        public static bool ReadYesNo(string question, bool required = false, bool defaultValue = false)
+        {
+            Console.WriteLine(question);
+            string choice = "";
+            bool isValid = false;
+            bool result = false;
+
+            if (required)
+            {
+                do
+                {
+                    choice = ReadInput(required);
+                    isValid = choice.Equals(Yes, StringComparison.CurrentCultureIgnoreCase) || choice.Equals(No, StringComparison.CurrentCultureIgnoreCase);
+
+                    if (isValid)
+                    {
+                        if (choice.Equals(Yes, StringComparison.CurrentCultureIgnoreCase))
+                        {
+                            result = true;
+                        }
+
+                        else if (choice.Equals(No, StringComparison.CurrentCultureIgnoreCase))
+                        {
+                            result = false;
+                        }
+                    }
+
+                    else
+                    {
+                        Console.WriteLine("Please, enter either \"yes\" or \"no\"");
+                    }
```

Comment:
You are checking the `choice` value twice for each possible input - lines 113, 117 and 122.
This is adding code duplication. You can either stick to what was in the pseudo code, or do something like this:
```csharp
bool isYes = choice.Equals(Yes);
bool isNo = choice.Equals(No);

isValid = isYes || isNo;
result = isYes;

if (!isValid)
{
    WriteLine("...");
}
```